src/plugins/data_explorer/public/components/sidebar/index.tsx

```diff
@@ -3,115 +3,95 @@
   const dispatch = useTypedDispatch();
-  const [options, setOptions] = useState<Array<EuiComboBoxOptionOption<string>>>([]);
-  const [selectedOption, setSelectedOption] = useState<EuiComboBoxOptionOption<string>>();
-  const { view, viewRegistry } = useView();
-  const views = viewRegistry.all();
-  const viewOptions = useMemo(
-    () =>
-      views.map(({ id, title }) => ({
-        value: id,
-        text: title,
-      })),
-    [views]
-  );
+  const [selectedSources, setSelectedSources] = useState<DataSourceOption[]>([]);
+  const [dataSourceOptionList, setDataSourceOptionList] = useState<DataSourceGroup[]>([]);
+  const [activeDataSources, setActiveDataSources] = useState<DataSourceType[]>([]);
 
   const {
     services: {
-      data: { indexPatterns },
+      data: { indexPatterns, dataSources },
       notifications: { toasts },
+      application,
     },
   } = useOpenSearchDashboards<DataExplorerServices>();
 
   useEffect(() => {
     let isMounted = true;
-    const fetchIndexPatterns = async () => {
-      await indexPatterns.ensureDefaultIndexPattern();
-      const cache = await indexPatterns.getCache();
-      const currentOptions = (cache || []).map((indexPattern) => ({
-        label: indexPattern.attributes.title,
-        value: indexPattern.id,
-      }));
-      if (isMounted) {
-        setOptions(currentOptions);
+    const subscription = dataSources.dataSourceService.dataSources$.subscribe(
+      (currentDataSources) => {
+        if (isMounted) {
+          setActiveDataSources(Object.values(currentDataSources));
+        }
       }
-    };
-    fetchIndexPatterns();
+    );
 
     return () => {
+      subscription.unsubscribe();
       isMounted = false;
     };
-  }, [indexPatterns]);
+  }, [indexPatterns, dataSources]);
+
+  const getMatchedOption = (dataSourceList: DataSourceGroup[], ipId: string) => {
+    for (const dsGroup of dataSourceList) {
+      const matchedOption = dsGroup.options.find((item) => item.value === ipId);
+      if (matchedOption !== undefined) return matchedOption;
+    }
+    return undefined;
+  };
 
-  // Set option to the current index pattern
   useEffect(() => {
     if (indexPatternId) {
-      const option = options.find((o) => o.value === indexPatternId);
-      setSelectedOption(option);
+      const option = getMatchedOption(dataSourceOptionList, indexPatternId);
+      setSelectedSources(option ? [option] : []);
+    }
+  }, [indexPatternId, activeDataSources, dataSourceOptionList]);
+
+  const handleSourceSelection = (selectedDataSources: DataSourceOption[]) => {
+    if (selectedDataSources.length === 0) {
+      setSelectedSources(selectedDataSources);
+      return;
     }
-  }, [indexPatternId, options]);
+    // Temporary redirection solution for 2.11, where clicking non-index-pattern datasource
+    // will redirect user to Observability event explorer
+    if (selectedDataSources[0]?.ds?.getType() !== 'DEFAULT_INDEX_PATTERNS') {
```

Comment:
`DEFAULT_INDEX_PATTERNS` should come from the `enum` that needs to be created or import from a common place.

Reply:
@mengweieric also the previous if guards against `selectedDataSources` being empty array and the class guarantees `getType` is defined, why are there optional chainings in this if?

Reply:
`selectedDataSources = [null]`

Reply:
sure, but it's `selectedDataSources: DataSourceOption[]` not `selectedDataSources: (DataSourceOption | null)[]`. 

if types (specially for objects created at the frontend, not server returned objects) are not reliable and we need to consider these issues happening at runtime, i think there are a lot more similar cases need the check, and the actual question would be how to avoid type inconsistency at runtime